moby/moby
# use terminfo in pkg/jsonmessage

**- What I did**

Switched `pkg/jsonmessage` to use the terminfo database for terminal control codes instead of hardcoding. This should be more portable to a wider variety of terminals etc

**- How I did it**

I vendored https://github.com/Nvveen/Gotty and ported `jsonmessage` to use it. For cases where no terminfo is available for $TERM or the terminfo entry is missing capabilities we use using I retained a fallback to the existing control codes.

I made a couple of improvements to the test's while doing this work and those are also included in this PR.

_concerns/caveats_:
* Surprisingly there is no obvious widely used/supported Go terminfo library. https://github.com/Nvveen/Gotty/ looks to be correct to me (and the best of the ones I looked at and is preferable to writing my own) but has not been touched for many years and only had one contributor (the original author).
* systems with broken terminfo may now misbehave where they may have gotten away with it before. systems with non-existent terminfo should see no change due to the fallbacks.
* the escape codes `jsonmessage` uses are pretty basic and well established ones, although as #28111 there is scope for differences even there (full-disclosure: terminfo would not actually have directly helped with #28111, see the fix in #28238), so it might be valid to say that this PR is not really warranted.

**- How to verify it**

Run a docker pull under a variety of terminals (I used rxvt-unicode, xterm, mlterm and aterm) and observe that the perceived output has not changed. Note that the actual control characters may now differ between terminals, this can be observed by using `script -c 'docker pull alpine' pull-alpine.script` and inspecting `pull-alpine.script`.

**- Description for the changelog**

Use terminfo database for control codes in docker pull et al.

**- A picture of a cute animal (not mandatory but encouraged)**
![Davy Jones, The Monkees](https://upload.wikimedia.org/wikipedia/commons/thumb/6/63/Davy_Jones%2C_Sydney_1968.jpg/440px-Davy_Jones%2C_Sydney_1968.jpg)
